### ISSUE TITLE
Idea for outputting files so that files are not overwritten with successive runs

### ISSUE THREAD

**Issue by juansc** [kdahlquist/GRNmap#144]
We are currently placing all the output files associated with a certain excel sheet in the same directory as the original sheet. I propose creating a new subdirectory for each run and storing all output files there.

Currently, if our directory is `mydir` and our spreadsheet is called mysheet.xlsx, after a run our directory will look like this:
- mydir
  - mysheet.xlsx
  - mysheet_output.xlsx
  - mysheet_output.mat
  - gene1.fig
  - gene2.fig

If we implement this new idea, the directory would look like this
- mydir
  - mysheet.xlsx
  - mysheet_output
    - mysheet_output.xlsx
    - mysheet_output.mat
    - gene1.fig
    - gene2.fig

I believe this idea would make the organization cleaner.

**Comment by kdahlquist** [comment]
Let's have some discussion about this at the meeting today.  I'd also like to hear what @bengfitzpatrick and @dondi say about this.

If we implement this, then you would still need to deal with the problem of successive runs over-writing files, but just at the directory level (issue #135).  Would you call the new directory "mysheet_output1", etc.?

**Comment by juansc** [comment]
Two other possible solutions:
* Have a save dialog to decide the directory
* There are some command line options to specify to save the output to a given location.

**Comment by kdahlquist** [comment]
I closed the redundant issue #135 Append successive numbers to the output files with the same names because it deals with the same issue as this one.

**Comment by kdahlquist** [comment]
If GRNmap detects a file with the same name, instead of overwriting it, append an underscore number, like "_1" or "_2", etc.

This is a necessary precursor to cueing up a series of input workbooks to run.

**Comment by kdahlquist** [comment]
If a user runs the same input workbook twice from the same directory, we don't want the new output to overwrite the previous run.

**Comment by dondi** [comment]
After some discussion, the specification has been revised so that:
* We will append a date-timestamp `yyyy-mm-dd-hh-mm-ss` to the base name _at all times_ to avoid naming collisions. The chance that two runs on the same input file ending at exactly the same second has been deemed highly unlikely.
* Date stamp can be based on time of writing of the output file.
* Check for the eventual filename length and display a message to the user, allowing them to either rename the file or have the program truncate the filename for them.
    * "Click Cancel to go back and change your filename. If you click OK, the program will run and truncate your filename."

**Comment by jtorre39** [comment]
@dondi Just confirming behaviour, by truncating we mean only truncating the resulting output file not the original input file name.

**Comment by dondi** [comment]
@jtorre39 had a question about determining filenames; we will use a constant to store the date format string and this will assure consistent file name length calculations.

An issue with tests that depend on filenames is anticipated, so we will see what happens in that case. @dondi thinks that such tests will probably need to be decoupled from reading/writings files.

**Comment by jtorre39** [comment]
This simple task is taking a surprising amount of time, to accomplish. This week all I was able to get done within the four hours was have the UI popup when a user picks an inputsheet that has too long of a name. Unfortunately, weighing the time it would take to test the UI in GRNmodel for this I felt would take too long as mocks in matlab don't seem too straightforward. 

Additionally, I wrote tests, and wrote a script that contains logic meant for truncating filenames that are too long, or if it finds a filename that is named appropriately it does nothing.

Next step is to create some tests / modify existing ones and then within Output.m hook in a call to my truncate script I just wrote and as long as all the tests pass this task can be closed.

**Comment by dondi** [comment]
Understood, some tasks do turn out that way. The good news is that you appear to be converging into closing this out. Any chance that a pull request can emerge from this by early next week?

**Comment by jtorre39** [comment]
Hopefully! The only thing that keeps me in doubt is that running the full test suite takes a few hours. Will do my best.

**Comment by jtorre39** [comment]
It's very difficult to depend on a filename for a test when the timestamp makes it unique per run. Our current setup makes it that we mandate having to run the 16 input sheets to generate updated expected files to test again as well.

**Comment by jtorre39** [comment]
I think ideally we refactor to saving output data into a struct, and then submitting that to a separate function that performs the file write as that seems far more testable. This will take quite a bit of time to do though especially since I would have to make expected output structs to test against.

**Comment by kdahlquist** [comment]
Based on your comments, I feel like we are overthinking this.  What we really want to do is come up with a mechanism for a user to avoid over-writing output files with successive runs.  Instead of appending the date/time stamp to every file, maybe we just need the first part, which is to check if there is an output file existing in the directory with the same name as the input file.  If there is, then give a message to the user that says that they are in danger of overwriting the results.  If they say, OK, then the run continues and over-writes the results.  Or they can cancel and copy/change the filename of the input file and start it again.

Will this solution get around some of the testing problems you identified, @jtorre39?

**Comment by jtorre39** [comment]
Yes, that would simplify a few testing problems actually. But there is the problem that our current tests for OutputTest.m don't test against an expected, it is actually just testing the actual against the actual. 

The current format for this test as well is difficult to maintain, as any changes for the output needs us to regenerate 16 new expected output files to test against. 

Refactoring the output function and tests can be made into a separate issue and we can take the path of least resistance for now to fix overwriting files.

**Comment by dondi** [comment]
It has been determined that just displaying a warning to the user regarding duplicate filenames will suffice, and the work to do that will be less than finishing the current path that @jtorre39 is on. So, he will (a) revert his work, (b) fix the discovered issues with the 16 file tests, then (c) implement the warning. He anticipates being finished by next week.

**Comment by dondi** [comment]
Warning dialog will state that there is a duplicate file. User may choose from the following:
* _OK_ —this will overwrite the duplicate file
* _Cancel_ —this will stop the process, with the expectation that the user does the needed file/naming adjustments before trying again

**Comment by jtorre39** [comment]
Commit up on my fork, running the entire test suite right now to make sure no regressions got in.

**Comment by jtorre39** [comment]
PR up.

**Comment by kdahlquist** [comment]
I'm not going to be at the meeting tomorrow, but it looks like with this issue being resolved, we are ready for another release.  I want to put the release off for a couple weeks so that @dondi and I can be in the same room to do it so that we can get Zenodo working for GRNmap.  Target date would be 10/18 (if I can come in for the meeting) or 10/25.

**Comment by kdahlquist** [comment]
There are a couple tweaks I'd like to make to this.
1. There is only an "OK" button.  There should be a "Cancel" button, too, especially since it is referred to in the message.
2. Can we change the message a little to say: "Warning: the output file that will be created for the chosen input file will have the same name as an existing output file.  Click OK to proceed with the model run, overwriting the existing output file. Click Cancel to stop the run so that you can rename the file.
3. The title text should read "Output File Name Conflict"

Thanks.

**Comment by kdahlquist** [comment]
Oops--besides there not being a cancel button, I tried to cancel by clicking the "x" in the upper-right corner.  However, this just closed the window, but did not cancel the run.  Clicking the "x" in the upper-right corner should have the same action as cancel, not "ok".

**Comment by jtorre39** [comment]
Message description updated and behavior fixed! 

Current behavior is now:
    - if user clicks "OK" the run will continue as normal
    - if user clicks "Cancel" or the X button or presses ESC on the keyboard the run will be cancelled

Changes are on my beta branch. Will be investigating a different issue to put them up together on a PR.

**Comment by kdahlquist** [comment]
I've confirmed the behavior on the beta branch of @jtorre39 's fork.  Once that is merged into beta, we can close this one.
